pycaw/callbacks.py

```diff
@@ -1,3 +1,101 @@
-# TODO
-# implement easy to use callbacks
-# (is already written, but 1 pull at the time)
+from ctypes import pointer
+
+from comtypes import COMObject
+
+from pycaw.api.audiopolicy import (IAudioSessionControl2, IAudioSessionEvents,
+                                   IAudioSessionNotification)
+from pycaw.api.endpointvolume import IAudioEndpointVolumeCallback
+from pycaw.utils import AudioSession
+
+
+class AudioSessionNotification(COMObject):
+    """
+    SESSION CREATED CALLBACK
+
+    """
+    _com_interfaces_ = [IAudioSessionNotification]
+
+    def OnSessionCreated(self, NewSession):
+        """Fired by Windows everytime a new session is created"""
+        ctl2 = NewSession.QueryInterface(IAudioSessionControl2)
+        new_session = AudioSession(ctl2)
+
+        self.on_session_created(new_session)
+
+    def on_session_created(self, new_session):
+        """pycaw user interface"""
+        pass
+
+
+class AudioSessionEvents(COMObject):
+    """Audio session callback"""
+    _com_interfaces_ = [IAudioSessionEvents]
+
+    # ======= DECODE RETURNED INT VALUE =======
+    # see audiosessiontypes.h and audiopolicy.h
+    AudioSessionState = (
+        "Inactive",
+        "Active",
+        "Expired"
+    )
+
+    AudioSessionDisconnectReason = (
+        "DeviceRemoval",
+        "ServerShutdown",
+        "FormatChanged",
+        "SessionLogoff",
+        "SessionDisconnected",
+        "ExclusiveModeOverride"
+    )
+
+    def OnSimpleVolumeChanged(self, new_volume, new_mute, event_context):
+        self.on_simple_volume_changed(new_volume, new_mute, event_context)
+
+    def OnStateChanged(self, new_state_id):
+        new_state = self.AudioSessionState[new_state_id]
+        self.on_state_changed(new_state, new_state_id)
+
+    def OnSessionDisconnected(self, disconnect_reason_id):
+        disconnect_reason = self.AudioSessionDisconnectReason[disconnect_reason_id]
+        self.on_session_disconnected(disconnect_reason, disconnect_reason_id)
+
+    # pycaw user interfaces:
+    def on_simple_volume_changed(self, new_volume, new_mute, event_context):
+        pass
+
+    def on_state_changed(self, new_state, new_state_id):
+        pass
+
+    def on_session_disconnected(self, disconnect_reason, disconnect_reason_id):
+        pass
+
+
+class AudioEndpointVolumeCallback(COMObject):
+    """Speaker callback"""
+    _com_interfaces_ = (IAudioEndpointVolumeCallback,)
+
+    def OnNotify(self, pNotify):
+
+        # get the data of the PAUDIO_VOLUME_NOTIFICATION_DATA Structure
+        _ = pNotify.contents
```

Comment:
If we're going to use it as a variable, we better name it something I'd say
```suggestion
        audio_volume_notification_data = pNotify.contents
```
Same for the other `_` variables below